client/basic-authorship/src/basic_authorship.rs

```diff
@@ -288,6 +311,8 @@ impl<A, B, Block, C> ProposerInner<B, Block, C, A>
 
 		let (block, storage_changes, proof) = block_builder.build()?.into_inner();
 
+		self.metrics.report(move |metrics| metrics.block_constructed.observe(started.elapsed().as_secs_f64()));
```

Comment:
I remember @mxinden having opinions regarding using the [HistogramTimer](https://docs.rs/prometheus/0.8.0/prometheus/struct.HistogramTimer.html) rather than `Instant::now() ... elapsed()` . Not sure that applies here though.

Reply:
Yes, I find the UX of the `HistogramTimer` API pretty neat, as dropping the timer automatically updates the histogram.

*(Very vague performance assumptions ahead)*

*Problem* is that it involves a [`clone`](https://docs.rs/prometheus/0.8.0/src/prometheus/histogram.rs.html#446) (Increase of the atomic reference counter) of the Histogram itself. While probably worth doing across function invocations, I don't think the additional CPU cycles (~100ns) are worth it within a single function. Even though maybe the compiler is able to optimize away the `clone`.

Thus :+1: for as it is.

Reply:
This is not very compatible with the fact that metrics may be off

Then we will have to do 

`let timer = self.metrics.map(|x| x.start_timere());`

and then

`if let Some(timer) = timer { .. }`

again

Reply:
You wouldn't need to do the latter (`if let Some(timer) = timer { .. }`) given that `HistogramTimer` autonomously `observe`s in its `Drop` implementation.

https://github.com/tikv/rust-prometheus/blob/master/src/histogram.rs#L400

But anyways :+1: for keeping it as it is.

Reply:
Ah right,

This function is rather heavy (it is executed once per block), so we can use this timer api actually

Reply:
@mxinden updated!